File: day10/haskell/part1.hs
import System.Environment (getArgs)
import Data.Bits (xor, testBit, popCount, shiftL)

-- Split string by delimiter
splitOn :: Char -> String -> [String]
splitOn _ [] = []
splitOn delim s = case break (== delim) s of
  (a, []) -> [a]
  (a, _:rest) -> a : splitOn delim rest

-- Extract all parenthesized groups from string
extractParens :: String -> [String]
extractParens [] = []
extractParens s = case dropWhile (/= '(') s of
  [] -> []
  (_:rest) -> takeWhile (/= ')') rest : extractParens (dropWhile (/= ')') rest)

-- Parse a line and return minimum presses
parseLine :: String -> Int
parseLine line = minPresses target buttons
  where
    -- Extract diagram between [ and ]
    diagram = takeWhile (/= ']') $ drop 1 $ dropWhile (/= '[') line
    numLights = length diagram
    target = sum [if c == '#' then 1 `shiftL` i else 0 | (i, c) <- zip [0..] diagram]

    -- Extract and parse buttons
    buttonStrs = extractParens line
    parseButton str = sum [1 `shiftL` idx | idxStr <- splitOn ',' str,
                                            let idx = read idxStr :: Int,
                                            idx < numLights]
    buttons = map parseButton buttonStrs

-- Find minimum presses for a machine
minPresses :: Int -> [Int] -> Int
minPresses target buttons = minimum validPresses
  where
    numButtons = length buttons
    allCombos = [0..2^numButtons - 1] :: [Int]

    compute combo = foldl xor 0 [buttons !! b | b <- [0..numButtons-1], testBit combo b]

    validPresses = [popCount combo | combo <- allCombos, compute combo == target]

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "input.txt" else head args
  content <- readFile filename
  let total = sum $ map parseLine $ lines content
  print total

File: day10/go/part1.go
package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}
	file, _ := os.Open(filename)
	defer file.Close()

	// Regex patterns
	diagramRe := regexp.MustCompile(`\[([.#]+)\]`)
	buttonRe := regexp.MustCompile(`\(([0-9,]+)\)`)

	totalPresses := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		// Parse diagram - target state
		diagramMatch := diagramRe.FindStringSubmatch(line)
		diagram := diagramMatch[1]
		numLights := len(diagram)

		// Target state as bitmask
		var target uint64
		for i, c := range diagram {
			if c == '#' {
				target |= (1 << i)
			}
		}

		// Parse buttons - each button is a bitmask of which lights it toggles
		buttonMatches := buttonRe.FindAllStringSubmatch(line, -1)
		var buttons []uint64
		for _, match := range buttonMatches {
			var buttonMask uint64
			indices := strings.Split(match[1], ",")
			for _, idx := range indices {
				i, _ := strconv.Atoi(idx)
				if i < numLights {
					buttonMask |= (1 << i)
				}
			}
			buttons = append(buttons, buttonMask)
		}

		// Try all combinations of button presses (2^numButtons)
		numButtons := len(buttons)
		minPresses := numButtons + 1 // worst case: impossible

		for combo := 0; combo < (1 << numButtons); combo++ {
			// Compute resulting state by XORing selected buttons
			var state uint64
			for b := 0; b < numButtons; b++ {
				if (combo>>b)&1 == 1 {
					state ^= buttons[b]
				}
			}

			if state == target {
				presses := bits.OnesCount(uint(combo))
				if presses < minPresses {
					minPresses = presses
				}
			}
		}

		totalPresses += minPresses
	}

	fmt.Println(totalPresses)
}
